File: backend/models/rfid_object.py
"""
models/rfid_object.py — Modello ORM per gli oggetti tracciati via RFID.

Tabella: rfid_objects

Ogni riga rappresenta un oggetto fisico (chiavi, ombrello, zaino, ecc.)
cui è stato associato un tag RFID UHF.
"""

import enum
from datetime import datetime

from sqlalchemy import Boolean, DateTime, Enum, Integer, String, func
from sqlalchemy.orm import Mapped, mapped_column

from database import Base


class ObjectStatus(str, enum.Enum):
    """
    Stato corrente dell'oggetto RFID.

    Valori:
        HOME:    l'oggetto si trova dentro casa (rilevato in ingresso o non uscito).
        AWAY:    l'oggetto è fuori casa (rilevato in uscita).
        UNKNOWN: stato non determinato (es. primo avvio, tag mai rilevato).
    """
    HOME = "home"
    AWAY = "away"
    UNKNOWN = "unknown"


class RfidObject(Base):
    """
    Oggetto fisico tracciato con tag RFID UHF.

    Colonne:
        id (int): PK auto-increment.
        name (str): Nome leggibile (es. "Chiavi di casa").
        rfid_uid (str): UID univoco del tag RFID (letto dal reader UHF).
        category (str): Categoria libera (es. "Keys", "Bag", "Documents").
        status (ObjectStatus): Stato corrente (home/away/unknown).
        is_sensitive (bool): Se True, genera alert quando portato fuori.
                             Utile per passaporti, medicinali, ecc.
        last_seen_at (datetime|None): Ultima volta che il tag è stato rilevato.
        created_at (datetime): Timestamp creazione record.
    """

    __tablename__ = "rfid_objects"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    name: Mapped[str] = mapped_column(String(150), nullable=False)
    # L'UID è la stringa esadecimale che il reader RFID UHF invia al Raspberry
    # Es. "E2000017221102141820B4E0" — deve essere univoco per ogni tag
    rfid_uid: Mapped[str] = mapped_column(String(64), unique=True, index=True, nullable=False)
    category: Mapped[str] = mapped_column(String(80), default="Generic", nullable=False)
    status: Mapped[ObjectStatus] = mapped_column(
        Enum(ObjectStatus), default=ObjectStatus.UNKNOWN, nullable=False
    )
    # Oggetti sensibili: generano notifica push anche senza bambini coinvolti
    is_sensitive: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    last_seen_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime, server_default=func.now(), nullable=False
    )

    def __repr__(self) -> str:
        return f"<RfidObject id={self.id} uid={self.rfid_uid!r} status={self.status}>"

File: app/lib/core/models/rfid_object.dart
// ============================================================
// RfidObject — modello dati per un oggetto taggato con RFID
// ============================================================
//
// Ogni oggetto fisico (chiavi, ombrello, zaino…) registrato
// nel sistema è rappresentato da questo modello.
// Il tag RFID (campo rfidTag) è l'ID univoco fisico dell'oggetto,
// ovvero l'EPC (Electronic Product Code) letto dal reader UHF.
//
// TODO: quando il backend è pronto, allineare i campi con
//       backend/app/models/schemas.py → RfidObjectSchema

/// Categoria dell'oggetto, usata per icone e filtri nell'UI.
enum ObjectCategory {
  keys,
  umbrella,
  bag,
  electronics,
  documents,
  clothing,
  other,
}

/// Stato corrente dell'oggetto rispetto alla porta di casa.
enum ObjectStatus {
  /// L'oggetto è dentro casa.
  inside,

  /// L'oggetto è uscito di casa.
  outside,

  /// Lo stato non è ancora noto (mai rilevato dal reader RFID).
  unknown,
}

/// Rappresenta un oggetto fisico tracciato via RFID nel sistema GateKeeper.
///
/// Ogni [RfidObject] è associato a un tag RFID UHF fisico.
/// Il Raspberry legge il tag quando l'oggetto passa vicino alla porta
/// e invia un evento al backend.
///
/// Esempio:
/// ```dart
/// final obj = RfidObject.fromJson(json);
/// if (obj.status == ObjectStatus.outside) {
///   // mostra notifica "hai dimenticato X"
/// }
/// ```
class RfidObject {
  /// ID univoco del tag RFID (EPC, formato es. "E200100860B20674").
  final String rfidTag;

  /// Nome leggibile assegnato dall'utente (es. "Chiavi di casa").
  final String name;

  /// Categoria per icona e filtraggio nell'UI.
  final ObjectCategory category;

  /// Stato attuale dell'oggetto (dentro/fuori/sconosciuto).
  final ObjectStatus status;

  /// ID dell'utente proprietario di questo oggetto.
  /// Null se l'oggetto non è assegnato a nessun utente specifico.
  // TODO: aggiungere UI per assegnare/riassegnare il proprietario
  final String? ownerId;

  /// True se l'app deve notificare quando questo oggetto esce senza
  /// che un utente adulto sia presente.
  final bool alertOnUnattendedExit;

  /// Timestamp dell'ultimo evento RFID rilevato per questo oggetto.
  final DateTime? lastSeenAt;

  const RfidObject({
    required this.rfidTag,
    required this.name,
    required this.category,
    required this.status,
    this.ownerId,
    required this.alertOnUnattendedExit,
    this.lastSeenAt,
  });

  /// Costruisce un [RfidObject] da JSON ricevuto dall'API.
  factory RfidObject.fromJson(Map<String, dynamic> json) {
    return RfidObject(
      rfidTag: json['rfid_tag'] as String,
      name: json['name'] as String,
      category: _categoryFromString(json['category'] as String),
      status: _statusFromString(json['status'] as String),
      ownerId: json['owner_id'] as String?,
      alertOnUnattendedExit:
          json['alert_on_unattended_exit'] as bool? ?? false,
      lastSeenAt: json['last_seen_at'] != null
          ? DateTime.parse(json['last_seen_at'] as String)
          : null,
    );
  }

  /// Serializza l'oggetto in Map da inviare all'API.
  Map<String, dynamic> toJson() {
    return {
      'rfid_tag': rfidTag,
      'name': name,
      'category': category.name,
      'status': status.name,
      if (ownerId != null) 'owner_id': ownerId,
      'alert_on_unattended_exit': alertOnUnattendedExit,
      if (lastSeenAt != null) 'last_seen_at': lastSeenAt!.toIso8601String(),
    };
  }

  /// Crea una copia con alcuni campi modificati.
  RfidObject copyWith({
    String? name,
    ObjectCategory? category,
    ObjectStatus? status,
    String? ownerId,
    bool? alertOnUnattendedExit,
    DateTime? lastSeenAt,
  }) {
    return RfidObject(
      rfidTag: rfidTag,
      name: name ?? this.name,
      category: category ?? this.category,
      status: status ?? this.status,
      ownerId: ownerId ?? this.ownerId,
      alertOnUnattendedExit:
          alertOnUnattendedExit ?? this.alertOnUnattendedExit,
      lastSeenAt: lastSeenAt ?? this.lastSeenAt,
    );
  }

  /// Restituisce l'emoji corrispondente alla categoria (usata nelle card UI).
  String get categoryEmoji {
    switch (category) {
      case ObjectCategory.keys:
        return '🔑';
      case ObjectCategory.umbrella:
        return '☂️';
      case ObjectCategory.bag:
        return '🎒';
      case ObjectCategory.electronics:
        return '💻';
      case ObjectCategory.documents:
        return '📄';
      case ObjectCategory.clothing:
        return '👕';
      case ObjectCategory.other:
        return '📦';
    }
  }

  static ObjectCategory _categoryFromString(String value) {
    return ObjectCategory.values.firstWhere(
      (e) => e.name == value,
      orElse: () => ObjectCategory.other,
    );
  }

  static ObjectStatus _statusFromString(String value) {
    return ObjectStatus.values.firstWhere(
      (e) => e.name == value,
      orElse: () => ObjectStatus.unknown,
    );
  }

  @override
  String toString() =>
      'RfidObject(tag: $rfidTag, name: $name, status: ${status.name})';

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      (other is RfidObject && other.rfidTag == rfidTag);

  @override
  int get hashCode => rfidTag.hashCode;
}

// ── Dati fake per lo sviluppo UI ──────────────────────────────────────
// TODO: rimuovere quando ApiService.getObjects() è connesso al backend.
const List<Map<String, dynamic>> kFakeObjectsJson = [
  {
    'rfid_tag': 'E200100860B20001',
    'name': 'Chiavi di casa',
    'category': 'keys',
    'status': 'inside',
    'owner_id': 'u-001',
    'alert_on_unattended_exit': true,
    'last_seen_at': '2026-04-29T18:30:00.000Z',
  },
  {
    'rfid_tag': 'E200100860B20002',
    'name': 'Ombrello nero',
    'category': 'umbrella',
    'status': 'outside',
    'owner_id': 'u-001',
    'alert_on_unattended_exit': false,
    'last_seen_at': '2026-04-29T08:15:00.000Z',
  },
  {
    'rfid_tag': 'E200100860B20003',
    'name': 'Zaino scuola',
    'category': 'bag',
    'status': 'inside',
    'owner_id': 'u-003',
    'alert_on_unattended_exit': true,
    'last_seen_at': '2026-04-28T17:00:00.000Z',
  },
  {
    'rfid_tag': 'E200100860B20004',
    'name': 'MacBook',
    'category': 'electronics',
    'status': 'outside',
    'owner_id': 'u-001',
    'alert_on_unattended_exit': false,
    'last_seen_at': '2026-04-29T09:00:00.000Z',
  },
];
